# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

In [1]:
! git clone https://www.github.com/DS3001/assignment2

Cloning into 'assignment2'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (36/36), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 36 (delta 9), reused 29 (delta 6), pack-reused 0
Receiving objects: 100% (36/36), 5.47 MiB | 28.56 MiB/s, done.
Resolving deltas: 100% (9/9), done.


In [2]:
import pandas as pd
import seaborn as sns
import numpy as np

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Question 1.1**

*This paper addresses the often-overlooked aspect of data cleaning known as "data tidying." It emphasizes the importance of structuring data into tidy datasets, where each variable corresponds to a column, each observation to a row, and each type of observational unit forms a separate table. Tidy datasets facilitate easier data manipulation, modeling, and visualization, and this framework simplifies dealing with messy datasets. Additionally, it highlights how adhering to this structured approach simplifies the development of tools for data analysis, both inputting and outputting tidy datasets. The paper illustrates the benefits of this approach through a case study, showcasing the efficiency gains in data analysis.*

**Question 1.2**

*The "tidy data standard" offers a structured and standardized framework for data organization, primarily aimed at streamlining and improving the data analysis process. It ensures consistency, reducing the need to start from scratch with each new dataset. Tidy data simplifies initial exploration, making data patterns and insights more accessible without extensive data wrangling. Moreover, it supports interoperability with "tidy tools" that handle both input and output of tidy datasets, reducing manual data translation between tools. This framework employs a user-friendly language, particularly beneficial for statisticians, aligning data structuring and analysis with their expertise. Ultimately, it empowers data analysts to focus on substantive analysis, minimizing the need for extensive data preparation and cleaning.*

**Question 1.3**

*In the context of data, it means that tidy datasets have a standardized and uniform structure, making them relatively similar and easy to work with. In contrast, messy datasets come in various disorganized forms, each with its unique challenges, making them difficult to handle.*

*The next sentence suggests that when you have a specific dataset in front of you, it's relatively straightforward to identify what constitutes an observation (individual data points or rows) and what constitutes a variable (attributes or columns). However, defining these terms in a general or abstract sense can be challenging. While it's clear within the context of a single dataset, creating a universally applicable definition that covers all datasets is a more complex task due to the diversity of data types, structures, and analytical needs in the field of data analysis.*

**Question 1.4**

*Values: Values are individual data points or measurements that are typically either numbers (quantitative) or strings (qualitative). In the context of your passage, values can represent specific information, such as "John" as a value for the variable "person" or "16" as a value for the variable "result." Values are the most granular level of data in a dataset.*

*Variables: Variables are attributes or characteristics that group and organize values that measure the same underlying concept or property across different units or observations. In your description, there are three variables: "person," "treatment," and "result." "Person" has three possible values (John, Mary, Jane), "treatment" has two possible values (a and b), and "result" has five or six values depending on how you handle the missing value.*

*Observations: Observations represent the individual units or entities on which data is collected. Each observation contains values for all variables. In your passage, you have a completely crossed experimental design where every combination of "person" and "treatment" was measured, resulting in six observations. An observation can be thought of as a specific person receiving a specific treatment and yielding a specific result.*

**Question 1.5**

*Tidy data is defined as a standardized way of organizing a dataset in which each variable corresponds to a column, each observation corresponds to a row, and each type of observational unit corresponds to a table, facilitating straightforward data analysis and extraction of variables.*

**Question 1.6**

*The five most common problems with messy datasets, as outlined in the passage, are as follows:*

*1. Column headers are values, not variable names: In this case, the data contains information in the column headers that should be part of the actual data, making it difficult to analyze.*

*2. Multiple variables are stored in one column: Data may be structured in a way where multiple variables are combined within a single column, making it challenging to extract and analyze these variables separately.*

*3. Variables are stored in both rows and columns: Variables may be scattered across both rows and columns, making it hard to distinguish between variables and observations.*

*4. Multiple types of observational units are stored in the same table: A single table may contain data from different types of observational units, leading to a lack of clear separation between them.*

*5. A single observational unit is stored in multiple tables: Data related to a single observational unit is divided into multiple tables, making it difficult to analyze the data holistically.*

*The data in Table 4 is messy because it violates the principles of tidy data. In this case, the issue is that multiple variables are stored in one column. The "table" column contains both the table names and the values that should be separate variables. This makes it challenging to work with the data in a straightforward manner.*

**"Melt" or "melting" a dataset refers to the process of reshaping it from a wide format (where variables are spread across columns) to a long format (where variables are stacked in a single column). This is typically done to make the data tidier and more amenable to analysis. Melting involves restructuring the data so that each row represents a unique observation, and each variable is stacked in a single column with a corresponding value column. This process helps resolve the issue of multiple variables stored in one column, making the data easier to work with.*

**Question 1.7**

*Table 11 is considered messy because it violates the principles of tidy data. The specific issues that make Table 11 messy include:*

*1. Variables are stored in both rows and columns: In Table 11, weather data is spread across rows (tmin, tmax) and columns (day, d1–d31). The temperature values (tmin and tmax) are found in the rows, while the day-specific values (d1–d31) are in the columns. This makes it challenging to work with and analyze the data efficiently.*

*2. Variable names are stored in a separate column (element): The element column in Table 11 stores the names of the variables (tmin and tmax) instead of having these variable names as individual columns. This makes the data structure less intuitive and violates the tidy data principle that each variable should form a column.*

*In Table 12, the data is tidy and "molten" because:*

*1. The data is "melted" because columns such as id, year, month, and element are identifier variables. Melting reshapes the data from a wide format (with variables in both rows and columns) into a long format, where each row represents a unique observation.*

*2. In Table 12(a), the missing values are dropped, which is permissible because the number of days in each month is known. This makes the missing values implicit rather than explicit, simplifying the presentation of the data.*

**Question 1.8**

*The "chicken-and-egg" problem in focusing on tidy data arises from its interdependence with tools designed for it, potentially hindering exploration of alternative data structures or manipulation methods. Wickham envisions future work in data wrangling to bring both incremental and substantial improvements. Tidy data, while a valuable framework, is not considered the final solution. He anticipates others will build on this foundation for better data storage and manipulation tools. Wickham also hopes that fields like user-centered design and human-computer interaction can contribute insights into the cognitive aspects of data analysis. Alternative data formulations, such as handling multidimensional arrays, and adaptive tools are potential areas of exploration. Addressing various data cleaning tasks, beyond just tidying data, is an ongoing endeavor to simplify and streamline the data preparation process.*

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

**Question 2.1**

In [ ]:
airbnb = pd.read_csv("assignment2/data/airbnb_hw.csv")
#print(airbnb)
#print(type(airbnb.Price))
airbnb['Price'].str.replace(',', '')
airbnb['Price'] = pd.to_numeric(airbnb['Price'], errors='coerce')


airbnb['Price'+'_nan'] = airbnb['Price'].isnull()


print('Total Missings: \n', sum(airbnb['Price'+'_nan']),'\n')


Total Missings: 
 181 



*The problem with the 'Price' variable in the Airbnb dataset was that once the price went about $1000, a comma was entered. This resulted in the variable being a character instead of numeric. The first part of this code replaced all of the commas with nothing. The next part coerces the variable into numeric and then changes all of the missing variables into NaN. In total there are about 181 missing variables out of about 30,000 observations*

**Question 2.2**

In [ ]:
sharksdf = pd.read_csv("assignment2/data/sharks.csv")

<ipython-input-4-6fa00c00b71b>:1: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sharksdf = pd.read_csv("assignment2/data/sharks.csv")


In [ ]:
print(sharksdf['Type'].unique(),'\n')

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg'] 



In [ ]:
sharksdf['Type'] = sharksdf['Type'].replace('Watercraft', 'Watercraft/Boat Accident')
sharksdf['Type'] = sharksdf['Type'].replace('Unverified', 'Unconfirmed')
sharksdf['Type'] = sharksdf['Type'].replace('Questionable', 'Unconfirmed')
sharksdf['Type'] = sharksdf['Type'].replace('Invalid', 'Unconfirmed')
sharksdf['Type'] = sharksdf['Type'].replace('Boating', 'Watercraft/Boat Accident')
sharksdf['Type'] = sharksdf['Type'].replace('Sea Disaster', 'Watercraft/Boat Accident')
sharksdf['Type'] = sharksdf['Type'].replace('Boat', 'Watercraft/Boat Accident')
sharksdf['Type'] = sharksdf['Type'].replace('Boatomg', 'Watercraft/Boat Accident')

In [ ]:
print(sharksdf['Type'].unique(),'\n')

['Unprovoked' 'Provoked' 'Unconfirmed' 'Watercraft/Boat Accident'
 'Under investigation' nan] 



*For this section, there were many observations that were either misspelled or that did not apply. Many of them were also very similar. For the ones that were similar like watercraft, boating, boat, and sea disaster I combined them all into one variable. I also combined the unverified, questionable, and invalid observations into one variable as well. This way the observations that have the most similarities are combined into one type.*

**Question 2.3**

In [ ]:
trial = pd.read_csv("VirginiaPretrialData2017.csv")

<ipython-input-5-d09c0302fe04>:1: DtypeWarning: Columns (1,4,5,7,79,80,81,82,83,84,108,163,164,165,166,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,301,302,303,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,3

In [ ]:
print(trial['WhetherDefendantWasReleasedPretrial'].unique(),'\n')

[9 0 1] 



In [ ]:
trial['WhetherDefendantWasReleasedPretrial'] = trial['WhetherDefendantWasReleasedPretrial'].replace(9, '') #replaces the 9's with blanks

trial['WhetherDefendantWasReleasedPretrial'] = trial['WhetherDefendantWasReleasedPretrial'].replace('',np.nan) #replaces the blanks with missing values


trial['WhetherDefendantWasReleasedPretrial'+'_nan'] = trial['WhetherDefendantWasReleasedPretrial'].isnull() #creates a variable for the missing values

print('Total Missings: \n', sum(trial['WhetherDefendantWasReleasedPretrial'+'_nan']),'\n')

Total Missings: 
 31 



In [ ]:
print(trial['WhetherDefendantWasReleasedPretrial'].unique(),'\n')

[nan  0.  1.] 



*For this question, I used the 'replace' function along with the np.nan to impute a missing value to all values that were listed as a 9. Then adding all of them, we see that there are 31 missing values for this variable*

**Question 2.4**

In [ ]:
print(trial['ImposedSentenceAllChargeInContactEvent'].unique(),'\n')

[' ' '60' '12' '.985626283367556' '36' '6' '24' '5.91375770020534' '120'
 '72' '11.9917864476386' '0' '2.95687885010267' '84' '108' '300' '240'
 '180' '4' '96' '2' '54' '.328542094455852' '44' '5' '115' '132' '48'
 '258' '34' '76' '.164271047227926' '.131416837782341' '111' '9' '3'
 '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '35' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '52' '30' '20'
 '192' '702' '14' '55' '53' '11.9055441478439' '114' '35.0061601642711'
 '68' '.657084188911704' '46.6242299794661' '102' '65' '200' '57'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '1500' '40' '284' '11' '118' '42' '162' '156'
 '47.2956878850103' '105' '51' '246' '29' '75' '324' '360'
 '34.4804928131417' '120.328542094456' '59.926078

In [ ]:
trial['ImposedSentenceAllChargeInContactEvent'] = pd.to_numeric(trial['ImposedSentenceAllChargeInContactEvent'], errors='coerce')

In [ ]:
trial['ImposedSentenceAllChargeInContactEvent'] = trial['ImposedSentenceAllChargeInContactEvent'].replace('',np.nan)

In [ ]:
trial['ImposedSentenceAllChargeInContactEvent'] = trial['ImposedSentenceAllChargeInContactEvent'] * 30.437

In [ ]:
trial['ImposedSentenceAllChargeInContactEvent'] = round(trial['ImposedSentenceAllChargeInContactEvent'])

*For this question, the variable was given in terms of months. Each month has a different number of days so that can lead to many and long decimals for each observation. To start, I made sure everything was numeric and then replaced all the blank values with np.nan's. From there I mulitplied the variable by 30.437 which is the average number of days per month. After that, I rounded the variable to the nearest whole number. The variable is now expressed in terms of whole days instead of months with long decimal places.*

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Question 5.1**

*The most recent US Census gathered data on race through a self-identification question, where individuals were asked to select their race or races from predefined categories, which included options like White, Black or African American, Asian, and more.*

**Question 5.2**

*Gathering demographic data, including data on race, serves several important purposes:*

1. *Resource Allocation: Data informs the distribution of government resources, like funding for schools, healthcare, and infrastructure, to address the needs of different communities.*

2. *Political Representation: Data determines the allocation of congressional seats, influencing political power and representation.*

3. *Civil Rights Enforcement: Data helps identify disparities and discrimination, aiding in the enforcement of civil rights laws.*

*Data quality is crucial because inaccurate data can lead to misallocation of resources, unequal representation, and hinder efforts to address social inequalities and disparities. Accurate data is essential for fair policy-making and social progress.*

**Question 5.3**

*The 2020 Census had several strengths, such as the introduction of online response, which was a significant step towards modernization. This made it more convenient for many individuals to participate and potentially reduced costs associated with paper-based surveys. The Census Bureau undertook extensive outreach and public awareness campaigns to encourage participation, increasing the overall response rate.*

*However, there were areas where the 2020 Census could have been improved.*

1. *Hard-to-Reach Populations: Despite the outreach efforts, certain hard-to-reach populations, like immigrants, minorities, and low-income individuals, may still be undercounted due to language barriers, mistrust, or fear. Future surveys should develop more targeted outreach strategies and provide resources to facilitate participation within these communities.*

2. *Digital Divide: While online response was a positive step, the digital divide remains a concern. Some people lack internet access or digital literacy, which could hinder their participation. Future surveys should ensure that all segments of the population have equal access to participation methods.*

3. *Data Privacy and Security: In an era of heightened concern about data privacy and security, the Census Bureau must continuously work to assure the public that their information is protected and their privacy respected.*

**Question 5.4**

*The 2020 Census collected data on sex through a binary male/female question, which has been the traditional approach. However, the Census did not collect specific data on gender identity, which is a significant limitation. Constructive criticism of this practice includes:*

1. *Lack of Inclusivity: The binary male/female question does not reflect the full spectrum of gender identities. It overlooks non-binary, genderqueer, transgender, and other gender-diverse individuals. Future surveys should include more inclusive gender options to respect and acknowledge the diversity of the population.*

2. *Data Relevance: Accurate and inclusive data on gender identity is crucial for policy-making, healthcare planning, and understanding disparities experienced by gender-diverse individuals. Without this data, it's challenging to address the unique needs of these communities effectively.*

3. *Data Privacy and Safety: It's essential to consider the privacy and safety of respondents when collecting gender identity data, as some individuals may fear discrimination or harm. Future practices should ensure anonymity and secure handling of this sensitive information.*

**Question 5.5**

*When cleaning data, especially data related to protected characteristics like sex, gender, sexual identity, or race, several concerns must be addressed. Firstly, there is a significant concern about data privacy and the potential for re-identification, as even anonymized data can sometimes be used to infer sensitive information. Ensuring robust data anonymization techniques and adherence to data protection regulations is crucial.
Secondly, when dealing with protected characteristics, missing values can be a challenge. The absence of data on these characteristics can lead to biased analysis and decision-making. Some common bad practices include ignoring missing values, which can perpetuate inequalities or misunderstand the true impact of policies on marginalized groups. Imputing missing data without considering the potential bias introduced is another concern, as it can lead to inaccurate results.
On the other hand, good practices involve careful handling of missing data, such as transparently reporting the extent of missing values and exploring the reasons behind these gaps. Collecting more comprehensive and inclusive data from the outset can also mitigate issues related to missing values. Additionally, conducting sensitivity analyses to assess the potential impact of missing data on results is a valuable practice. The key is to be aware of the unique challenges associated with protected characteristics and to approach data cleaning with a commitment to fairness, accuracy, and privacy.*

**Question 5.6**

*Imputing protected characteristics like race, gender, sex, or sexuality through algorithms raises ethical concerns regarding individual autonomy and privacy, potentially perpetuating stereotypes and biases. It risks infringing upon the right to self-identify. Ensuring fairness and accuracy is vital, as algorithmic biases could exacerbate societal prejudices, impacting decision-making and policy formulation. The fluid nature of these characteristics and the quality of initial data further complicate the accuracy of imputed values. Additionally, legal and regulatory compliance, particularly in the context of privacy laws, must be addressed when handling such sensitive data. Imputing these characteristics should be approached with extreme caution, with a focus on transparency, consent, privacy protection, and the prevention of biases, while adhering to relevant legal mandates.*